### PR TITLE
Fix/benchmark messages memory usage

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

Perform various duration tests to  check memory usage across the time.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: benchmark/checks/check_mem_usage.py [hunk 1]
```diff
@@ -57,6 +57,10 @@ def handle(self, message: Message) -> None:
 
 def run(duration, runtime_mode):
     """Check memory usage."""
+    # pylint: disable=import-outside-toplevel
+    # import causer tricky import in decision_maker
```
Comment (DavidMinarsch) [inline benchmark/checks/check_mem_usage.py:61]:
```suggestion
    # import manually due to some lazy imports in decision_maker
```

File: benchmark/checks/check_multiagent.py [hunk 2]
```diff
@@ -105,6 +105,10 @@ def handle(self, message: Message) -> None:
 
 def run(duration, runtime_mode, runner_mode, start_messages, num_of_agents):
     """Test multiagent message exchange."""
+    # pylint: disable=import-outside-toplevel
+    # import causer tricky import in decision_maker
```
Comment (DavidMinarsch) [inline benchmark/checks/check_multiagent.py:109]: see above

File: benchmark/checks/check_reactive.py [hunk 3]
```diff
@@ -84,6 +84,10 @@ def handle(self, message: Message) -> None:
 
 def run(duration, runtime_mode, connection_mode):
     """Test memory usage."""
+    # pylint: disable=import-outside-toplevel
+    # import causer tricky import in decision_maker
```
Comment (DavidMinarsch) [inline benchmark/checks/check_reactive.py:88]: see above

File: benchmark/checks/check_proactive.py [hunk 4]
```diff
@@ -61,6 +61,10 @@ def act(self):
 
 def run(duration, runtime_mode):
     """Test act message generate performance."""
+    # pylint: disable=import-outside-toplevel
+    # import causer tricky import in decision_maker
```
Comment (DavidMinarsch) [inline benchmark/checks/check_proactive.py:65]: see above

File: benchmark/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+ENV PYTHONPATH "$PYTHONPATH:/usr/lib/python3.8/site-packages"
+RUN apk add --no-cache make git bash wget
+RUN apk add --no-cache gcc musl-dev python3-dev libffi-dev openssl-dev
+RUN apk add --update --no-cache py3-numpy py3-scipy py3-pillow
+RUN apk add --update --no-cache gfortran freetype-dev libpng-dev openblas-dev g++ py3-numpy-dev
+RUN apk add --no-cache go
+
+RUN pip install --upgrade pip pipenv
+RUN pip install protobuf colorlog graphviz
```
Comment (DavidMinarsch) [inline benchmark/Dockerfile:11]: I don't think we need colorlog graphviz any more. It was needed for the old oef. Please check

File: benchmark/checks/check_dialogues_memory_usage.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    @property
+    def random_string(self) -> str:
+        """Get random string on every access."""
+        return uuid.uuid4().hex
+
+    def process_message(self) -> None:
+        """Process a message with dialogues."""
+        message = self.create()
+        dialogue = self.update(message)
+        self.reply(dialogue, message)
```
Comment (DavidMinarsch) [inline benchmark/checks/check_dialogues_memory_usage.py:65]: This is wrong. The dialogues model has a `create` method. use this, see the skills for examples!

Comment (DavidMinarsch) [reply benchmark/checks/check_dialogues_memory_usage.py:65]: e.g. generic buyer skill, behaviours.py line 81

File: benchmark/checks/run_benchmark_messages_mem.sh [hunk 7]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+DATE=`LC_ALL=C date +"%d.%m.%Y_%H:%M"`
+echo "Performance report for $DATE"
+echo -e "-----------------------------\n"
+
+AGENTS=10
+NUM_RUNS=10
```
Comment (DavidMinarsch) [inline benchmark/checks/run_benchmark_messages_mem.sh:7]: Number runs should be higher to get more reliable measures. I'd say 100

File: benchmark/checks/check_dialogues_memory_usage.py (hunk 6 above)
Comment (solarw) [reply benchmark/checks/check_dialogues_memory_usage.py:65]: every message generated is sent from a new address.
so the message has to be generated manually,  existing dialogues can not be used, cause the same dialogue can not be used for creating and reply (checked!)

File: benchmark/checks/run_benchmark_messages_mem.sh (hunk 7 above)
Comment (solarw) [reply benchmark/checks/run_benchmark_messages_mem.sh:7]: will update a bit later

File: benchmark/checks/check_multiagent.py (hunk 2 above)
Comment (solarw) [reply benchmark/checks/check_multiagent.py]: done

File: benchmark/checks/check_mem_usage.py (hunk 1 above)
Comment (solarw) [reply benchmark/checks/check_mem_usage.py]: done

File: benchmark/checks/check_proactive.py (hunk 4 above)
Comment (solarw) [reply benchmark/checks/check_proactive.py]: done

File: benchmark/checks/check_reactive.py (hunk 3 above)
Comment (solarw) [reply benchmark/checks/check_reactive.py]: done

File: benchmark/Dockerfile (hunk 5 above)
Comment (solarw) [reply benchmark/Dockerfile]: fixed

File: benchmark/checks/check_dialogues_memory_usage.py (hunk 6 above)
Comment (DavidMinarsch) [reply benchmark/checks/check_dialogues_memory_usage.py:65]: Yes you'd need to create two dialogues. That's better though as it keeps it consistent. The message should be created outside the handler (e.g. in a behaviour) and just then processed by the handler. Your custom logic here does not reflect what we have in the framework.

Comment (solarw) [reply benchmark/checks/check_dialogues_memory_usage.py:65]: realistic case implemented in `check_multiagent_http_dialogues.py`, a limited amount of agents are interacting, there is one agent interacting with like an unlimited amount of agent, so every message generated from the new address

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**